File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    
    dic1 = {} # {장르: 총 재생 횟수}
    dic2 = {} # {장르: [(플레이 횟수, 고유번호)]}
    
    for i in range(len(genres)):
        dic1[genres[i]] = dic1.get(genres[i], 0) + plays[i]
        dic2[genres[i]] = dic2.get(genres[i], []) + [(plays[i] , i)]
    
    # 재생 횟수 내림차순으로 장르별 정렬
    dic2_sort = sorted(dic1.items(), key=lambda x: x[1], reverse=True)
    
    # 재생 횟수 내림차순, 인덱스 오름차순 정렬
    for (genre, play) in dic2_sort:
        dic2[genre] = sorted(dic2[genre], key=lambda x : (-x[0], x[1]))
        answer += [i for (play, i) in dic2[genre][:2]]
        
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        // 장르별 총 재생 횟수를 저장하는 Map
        HashMap<String, Integer> genrePlayCount = new HashMap<>();
        // 장르별 (플레이 횟수, 고유번호)의 리스트를 저장하는 Map
        HashMap<String, List<int[]>> genreSongs = new HashMap<>();
        
        for (int i = 0; i < genres.length; i++) {
            String genre = genres[i];
            int play = plays[i];
            
            // 장르별 총 재생 횟수 저장
            genrePlayCount.put(genre, genrePlayCount.getOrDefault(genre, 0) + play);
            
            // 장르별 곡의 (재생 횟수, 고유번호) 저장
            genreSongs.putIfAbsent(genre, new ArrayList<>());
            genreSongs.get(genre).add(new int[] {play, i});
        }
        
        // 장르별 총 재생 횟수 기준으로 내림차순 정렬
        List<String> genreSort = new ArrayList<>(genrePlayCount.keySet());
        genreSort.sort((a, b) -> genrePlayCount.get(b) - genrePlayCount.get(a));
        
        List<Integer> answer = new ArrayList<>();
        
        // 각 장르별로 곡들을 재생 횟수 내림차순, 고유번호 오름차순으로 정렬 후 상위 2개 추출
        for (String genre : genreSort) {
            List<int[]> songs = genreSongs.get(genre);
            
            // 재생 횟수 내림차순, 고유번호 오름차순 정렬
            songs.sort((a, b) -> {
                if (b[0] == a[0]) {
                    return a[1] - b[1]; // 재생 횟수가 같으면 고유번호 오름차순
                } else {
                    return b[0] - a[0]; // 재생 횟수 내림차순
                }
            });
            
            // 상위 2개 곡 추가
            for (int i = 0; i < Math.min(2, songs.size()); i++) {
                answer.add(songs.get(i)[1]); // 고유번호를 리스트에 추가
            }
        }
        
        // 정답 리스트를 배열로 변환
        return answer.stream().mapToInt(i -> i).toArray();
    }
}
